refaster-runner/src/main/java/tech/picnic/errorprone/refaster/runner/RefasterRuleSelector.java

```diff
@@ -0,0 +1,554 @@
+import com.google.errorprone.refaster.UExpression;
+import com.google.errorprone.refaster.UStatement;
+import com.google.errorprone.refaster.UStaticIdent;
+import com.google.errorprone.refaster.annotation.BeforeTemplate;
+import com.sun.source.tree.AssignmentTree;
+import com.sun.source.tree.BinaryTree;
+import com.sun.source.tree.ClassTree;
+import com.sun.source.tree.CompilationUnitTree;
+import com.sun.source.tree.CompoundAssignmentTree;
+import com.sun.source.tree.ExpressionTree;
+import com.sun.source.tree.IdentifierTree;
+import com.sun.source.tree.MemberReferenceTree;
+import com.sun.source.tree.MemberSelectTree;
+import com.sun.source.tree.MethodTree;
+import com.sun.source.tree.PackageTree;
+import com.sun.source.tree.Tree;
+import com.sun.source.tree.UnaryTree;
+import com.sun.source.tree.VariableTree;
+import com.sun.source.util.TreeScanner;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.IdentityHashMap;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Consumer;
+import javax.annotation.Nullable;
+
+// XXX: Add some examples of which source files would match what templates in the tree.
+// XXX: Consider this text in general.
+/**
+ * A {@link RefasterRuleSelector} algorithm that selects Refaster templates based on the content of
+ * a {@link CompilationUnitTree}.
+ *
+ * <p>The algorithm consists of the following steps:
+ *
+ * <ol>
+ *   <li>Create a {@link Node tree} structure based on the provided Refaster templates.
+ *       <ol>
+ *         <li>Extract all identifiers from the {@link BeforeTemplate}s.
+ *         <li>Sort identifiers lexicographically and collect into a set.
+ *         <li>Add a path to the tree based on the sorted identifiers.
+ *       </ol>
+ *   <li>Extract all identifiers from the {@link CompilationUnitTree} and sort them
+ *       lexicographically.
+ *   <li>Traverse the tree based on the identifiers from the {@link CompilationUnitTree}. Every node
+ *       can contain Refaster templates. Once a node is we found a candidate Refaster template and
+ *       will therefore be added to the list of candidates.
+ * </ol>
+ *
+ * <p>This is an example to explain the algorithm. Consider the templates with identifiers; {@code
+ * T1 = [A, B, C]}, {@code T2 = [B]}, and {@code T3 = [B, D]}. This will result in the following
+ * tree structure:
+ *
+ * <pre>{@code
+ * <root>
+ *    ├── A
+ *    │   └── B
+ *    │       └── C -- T1
+ *    └── B         -- T2
+ *        └── D     -- T3
+ * }</pre>
+ *
+ * <p>The tree is traversed based on the identifiers in the {@link CompilationUnitTree}. When a leaf
+ * contains a template and is reached, we can be certain that the identifiers from the {@link
+ * BeforeTemplate} are at least present in the {@link CompilationUnitTree}.
+ *
+ * <p>Since the identifiers are sorted, we can prune parts of the {@link Node tree} while we are
```

Comment:
Prune implies modification of the tree.
```suggestion
 * <p>Since the identifiers are sorted, we can skip parts of the {@link Node tree} while we are
```